src/basic/efivars.c

```diff
@@ -271,11 +277,21 @@ static int read_flag(const char *varname) {
 }
 
 bool is_efi_secure_boot(void) {
-        return read_flag("SecureBoot") > 0;
+        static int cache = -1;
+
+        if (cache < 0)
+                cache = read_flag("SecureBoot");
+
+        return cache > 0;
 }
 
 bool is_efi_secure_boot_setup_mode(void) {
-        return read_flag("SetupMode") > 0;
+        static int cache = -1;
+
+        if (cache < 0)
+                cache = read_flag("SetupMode");
+
+        return cache > 0;
```

Comment:
Don't we also want to cache access to `SystemOptions`? (the function below this one)

Reply:
hmm, we access that a lot less often, no?

caching a string is much more complex, because we don't want to leak data, and have to deal with concurrency then. i.e. we generally write our code so that it works reasonably in a threaded environment, even if we try to avoid threads. If our cache is just an "int" we don't need to do much for that, as on all systems we know their are atomic, i.e. either the cache is seen uninitialized or initialized, and in the worst case we see it from multiple threads uninitialized, which means we initialize it twice, but that doesn't matter. for strings we'd start to have using CAS then however, in order not to leak the string we place there, and similar.

Hence, unless the SystemdOptions EFI var is really accessed that often, too I don't think we need to do anything here.

or to say this differently: it's probably easier to cache what we parse out of SystemdOptions than the whole strings.

Reply:
Given that my patch that 'triggered' this issue puts `is_efi_secure_boot` in the same codepath as the  `SystemdOptions`  variable read, (namely whether to read that variable is conditional on the fact that `is_efi_secure_boot` is true)   I would argue it is (at least for that usecase) read just as often.

Reply:
I'm not sure. proc_cmdline_parse would normally be invoked just once per process lifetime.
@poettering, did you do any profiling of how much this patch resolves the issue?

Anyway, I think this is good to merge. It might resolve the issue entirely (if we happen to stay below the per-process limit), or it might solve it partially.

Reply:
@keszybz i did not do any profiling...